src/lib.rs

```diff
@@ -0,0 +1,289 @@
+                return -1;
+            }
+        }
+    };
+    let write_buf: &[u8] = unsafe {
+        if buf.is_null() {
+            eprintln!("ClientSession::write: buf was NULL");
+            return -1;
+        }
+        slice::from_raw_parts(buf, count as usize)
+    };
+    let n_written: usize = match session.write(write_buf) {
+        Ok(n) => n,
+        Err(e) => {
+            eprintln!("ClientSession::write: {}", e);
+            return -1;
+        }
+    };
+    n_written as ssize_t
+}
+
+// Read plaintext bytes from the ClientSession. This acts like
+// read(2). It returns the number of bytes read, or -1 on error.
+#[no_mangle]
+pub extern "C" fn rustls_client_session_read(
+    session: *const c_void,
+    buf: *mut u8,
+    count: size_t,
+) -> ssize_t {
+    let session: &mut ClientSession = unsafe {
+        match (session as *mut ClientSession).as_mut() {
+            Some(cs) => cs,
+            None => {
+                eprintln!("ClientSession::process_new_packets: session was NULL");
+                return -1;
+            }
+        }
+    };
+    let read_buf: &mut [u8] = unsafe {
+        if buf.is_null() {
+            eprintln!("ClientSession::read: buf was NULL");
+            return -1;
+        }
+        slice::from_raw_parts_mut(buf, count as usize)
+    };
+    let n_read: usize = match session.read(read_buf) {
+        Ok(n) => n,
+        // The CloseNotify TLS alert is benign, but rustls returns it as an Error. See comment on
+        // https://docs.rs/rustls/0.19.0/rustls/struct.ClientSession.html#impl-Read.
+        // Log it and return EOF.
+        Err(e) if e.kind() == ConnectionAborted && e.to_string().contains("CloseNotify") => {
+            eprintln!(
+                "ClientSession::read_tls: CloseNotify (this is expected): {}",
+                e
+            );
+            return 0;
+        }
+        Err(e) => {
+            eprintln!("ClientSession::read: {}", e);
+            return -1;
+        }
+    };
+    n_read as ssize_t
+}
+
+// Read TLS bytes taken from a socket into the ClientSession. This acts like
+// read(2). It returns the number of bytes read, or -1 on error.
+#[no_mangle]
+pub extern "C" fn rustls_client_session_read_tls(
+    session: *const c_void,
+    buf: *const u8,
+    count: size_t,
+) -> ssize_t {
+    let session: &mut ClientSession = unsafe {
+        match (session as *mut ClientSession).as_mut() {
+            Some(cs) => cs,
+            None => {
+                eprintln!("ClientSession::process_new_packets: session was NULL");
+                return -1;
+            }
+        }
+    };
+    let input_buf: &[u8] = unsafe {
+        if buf.is_null() {
+            eprintln!("ClientSession::read_tls: buf was NULL");
+            return -1;
+        }
+        slice::from_raw_parts(buf, count as usize)
+    };
+    let mut cursor = Cursor::new(input_buf);
+    let n_read: usize = match session.read_tls(&mut cursor) {
```

Comment:
It's not clear to me that this is sound -- the caller is quite likely to provide a `buf` which contains uninitialized memory, I believe turning those until a slice is technically not sound.

Reply:
Any tips on where I can read about the soundness of turning uninitialized memory into a slice? Do I just need to initialize it? Before or after `from_raw_parts`?

Reply:
I don't have a great citation (boo me!), it's only a soundness issue if `read_tls` ever _reads_ from `&mut cursor`, as long as it writes to it only, then it's find.

This is, classically, a big issue in the design of `io::Read`, which led to this whole thing https://github.com/rust-lang/rust/pull/42002